libbpf-tools/numamove.bpf.c

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "numamove.h"
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, u32);
+	__type(value, u64);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} start SEC(".maps");
+
+static struct data data;
+
+SEC("fentry/migrate_misplaced_page")
+int BPF_PROG(migrate_misplaced_page)
+{
+	u32 pid = bpf_get_current_pid_tgid();
+	u64 ts = bpf_ktime_get_ns();
+
+	bpf_map_update_elem(&start, &pid, &ts, 0);
+	return 0;
+}
+
+SEC("fexit/migrate_misplaced_page")
+int BPF_PROG(migrate_misplaced_page_exit)
+{
+	u32 pid = bpf_get_current_pid_tgid();
+	u64 *tsp, ts = bpf_ktime_get_ns();
+	s64 delta;
+
+	tsp = bpf_map_lookup_elem(&start, &pid);
+	if (!tsp)
+		return 0;
+	delta = (s64)(ts - *tsp);
+	if (delta < 0)
+		goto cleanup;
+	__sync_fetch_and_add(&data.lat, delta / 1000000U);
+	__sync_fetch_and_add(&data.num, 1);
+
+cleanup:
+	bpf_map_delete_elem(&start, &pid);
```

Comment:
Yes, `migrate_misplaced_page` occurs in abnormal mode, so its processing is interruptible and allowed to sleep and schedule.
I used a heavy load test and found that it would change.